feat(RAIN-39306) CLI policy upload/download support

satellite15 opened the pull request:
On upload we create the root policy dirs 'policies/{module}' and write these as tar entries
  to ensure we set permissions and mod time.

  On upload we also do the policyId/sid rewrite

  Download now downloads a tar of tars and extracts these to the users latest policies dir.
  These changes extract those tars within that dir combining customer and lacework policies.
  When running a scan we do not rewrite the policyId/sid, these have been validated and corrected
  by the api server and the metadata.yaml contains the values that should be used in a scan.

Signed-off-by: Martin Scott <martin.scott@lacework.net>

slshen commented on pkg/xcp/xcp.go:
```diff
@@ -211,6 +212,7 @@ func GetCIEnv(dir string) map[string]string {
 		values["SOLUBLE_METADATA_HOSTNAME"] = h
 	}
 
+	values["POLICY_METADATA_UPLOAD_TIME"] = time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
```
We usually record this on the server side instead of having the CLI provide it.

Also, this gets runs for every assessment upload, so if we really want to include this, and it's fine if we do, then it needs to be renamed.  Maybe simply UPLOAD_TIME.

There's a page on the internal wiki listing the upload variables and they're meaning.  Make sure to update that as well.

slshen commented on pkg/policy/policy.go:
```diff
@@ -49,6 +49,7 @@ const (
 	Secrets        = Target("secrets")
 	ARM            = Target("arm")
 	None           = Target("")
+	Policies       = "policies"
```
I have avoided constants like this, I find it's better to just write the literal string "policies" rather than policy.Policies.

slshen commented on pkg/policy/policy.go:
```diff
@@ -71,9 +72,23 @@ type PolicyType interface {
+	Dir                    string
+	Policies               map[PolicyType][]*Policy
+	PolicyIds              map[string]string
+	SkipPolicyIDResolution bool
```
Can you explain the need to make this optional?  I can't find where it's used in the code.

Also, I suggest using the name "validation" instead of "resolution".

satellite15 replied:
When running a scan the downloaded policies will have sid/policyId that have been rewritten by the api-server for lacework policies and all custom policies will be vetted to start `c-` as per the design. Thus we don't want the cli building the policyId from the folder name when running a scan or we end up with `c-` policies again for lacework policies. It's set when creating the store after downloading the policy archive and It's used when loading policies 

```

	if m.SkipPolicyIDResolution {
		policy.ID = policy.Metadata.GetString("policyId")
	} else {
		id, err := m.resolvePolicyID(policyType, path)
		if err != nil {
			return nil, err
		}
		policy.ID = id
		policy.Metadata["policyId"] = policy.ID
		policy.Metadata["sid"] = policy.ID
	}
```

satellite15 commented:
API is reviewed and merged so merging these changes to ensure CLI has what's needed to continue with Opal phase 1. Any issues to be addressed as bugs in JIRA. Addressing comments in new PR.
